### PR TITLE
Made Button a container, solves #290

### PR DESCRIPTION
**Pull request by tkrugg**
- made Button a Container
- removed the part that trims innerText to fill in `label`
- set containerNode before the icon and the label
- I updated the samples (they were setting label using innerText) and added an example with a ProgressIndicator inside a ToggleButton.
- updated doc

> Quoted from doc: The only string displayed by the button is the one defined by its `label` property.

I took the liberty of removing this part of the doc, i think it's no longer relevant.

### CODE REVIEW

**Comment by wkeese** [comment]
Hmm, this doesn't seem right.   You have two separate nodes: `containerNode` and `labelNode`.  Shouldn't it just be one node?   Because, for example, `myButton.label = "OK"` should clear the button text set originally in markup, ex: `<d-button>hello</d-button>`.

**Comment by tkrugg** [comment]
`.label` doesn't seem important in Button but it needs to be kept because it's used in ToggleButton (used along with `.checkedLabel`), and there are situations where we might want to have both, as to make the following kind of refresh button, embedding a ProgressIndicator:

image

Also currently, it's impossible to get an `<icon><text>` kind of button without a label.

**Comment by wkeese** [comment]
> there are situations where we might want to have both, as to make the following kind of refresh button, embedding a ProgressIndicator

If a user wants a button with a label and a ProgressIndicator, he can easily do:

```html
<d-button>
   <d-progress-indicator></d-progress-indicator>
   refreshing...
</d-button>
```


> Also currently, it's impossible to get an <icon><text> kind of button with containerNode only.

That's not true, just check out http://download.dojotoolkit.org/release-1.10.2/dojo-release-1.10.2/dijit/tests/form/test_Button.html.  It has many examples like that.

**Comment by cjolif** [comment]
What is missing is enhancing existing tests to tests for that new feature?

**Comment by cjolif** [comment]
I personally have no issue having the label not clearing the content. Actually I think it is probably easier for the user to understand, there will be not trick of setting a particular property influencing the behavior.

**Comment by tkrugg** [comment]
> That's not true, just check out http://download.dojotoolkit.org/release-1.10.2/dojo-release-1.10.2/dijit/tests/form/test_Button.html. It has many examples like that.

By "currently" I meant in this PR. But with the dijit samples I understand better the bahviour you described in https://github.com/ibm-js/deliteful/pull/513#issuecomment-72966489. I'll implement that.

**Comment by wkeese** [comment]
> I personally have no issue having the label not clearing the content. Actually I think it is probably easier for the user to understand, there will be not trick of setting a particular property influencing the behavior.

Well, I think the behavior would be counterintuitive to users.  The common use cases of buttons are:

```html
<d-button id=myButton>OK</d-button>
```

or creating programatically:

```js
myButton = new Button({label: "OK});
```

99% of users (especially HTML programmers) will see these two as identical, using the terms "content", "text" and "label" interchangeably.   In other words, if the application later does:

```js
myButton.label = "Cancel";
```

Then I think users will expect the text displayed by the button to be "Cancel", not "OK Cancel" or "Cancel OK".

**Comment by cjolif** [comment]
What you say seems to imply that label property must be treated differently than other properties so that it is easier for users. What I'm saying is that users find thing easier when they are consistent (i.e. in this case label is no different than other properties). I guess both are valid ways of seeing it. I'm a bit frighten that you find the first one more obvious because we are used to do that in Dijit. But that does not necessarily mean all users would see it like that.

**Comment by edchat** [comment]
cjolif, the first example I see when looking up html5 button is:
 `<button type="button">Click Me!</button> ` so it seems like that is common.

**Comment by cjolif** [comment]
@edchat so what, this will work?

**Comment by edchat** [comment]
@cjolif when you said "I'm a bit frighten that you find the first one more obvious because we are used to do that in Dijit." I thought you meant that syntax.  Is that not the first one you were talking about?

**Comment by cjolif** [comment]
> when you said "I'm a bit frighten that you find the first one more obvious because we are used to do that in Dijit." I thought you meant that syntax. Is that not the first one you were talking about?

I meant that `<button label="foo">bar</button>` having label "foo" overriding "bar" being dijit behavior (first solution) comparing to @tkrugg initial PR which was displaying both (2nd solution).

**Comment by wkeese** [comment]
As @edchat suggested, I'm trying to match the behavior on native HTML, not trying to match dijit.   Also, `<button label="foo">bar</button>` is a weird example because no one would ever do that.

**Comment by cjolif** [comment]
> As @edchat suggested, I'm trying to match the behavior of native HTML, not trying to match dijit.

I'm probably missing something but what in @tkrugg solution is not matching native HTML behavior? If you don't put a label property (which the native element does not have) you get the same visual rendering as the native button, aren't you? The only case where you have a different behavior is if you use a label property, but that's expected as anyway native button does not have a label property...

I guess I'll come back to my initial point which is why making label such a specific case? We have not other property fo the widget that is filled automatically by its content? Also if we do that we should also make sure to listen to changes of the content to re-apply it correctly?  (which it is not doing today?)

In other words doing that 100% correctly (or even 95%) is it seems to me quite some code and I don't really see the necessity of "paying" for it? We are not losing functionality without it? (i.e. you want to use label, you use it, you want to use child content you use it, but what is the thing we are losing by not replicating child content into label?)

**Comment by wkeese** [comment]
> what is the thing we are losing by not replicating child content into label?

As I said above, having both "label" and "content" as distinct properties leads to unexpected behavior and makes it slightly difficult for the application to programatically change the text of a button that was created declaratively.

**Comment by cjolif** [comment]
> makes it slightly difficult for the application to programatically change the text of a button that was created declaratively.

There are two ways to setup the button programmatically, if you actually set up a label property it should not be hard to change its label? 

You are right in the case you don't use the label property but the "content". But in that case you are not worst than a pure HTML button, or are you? If you are then I agree this is problematic, if not one more time I'm ready to live with this for the sake of simpler code.

**Comment by wkeese** [comment]
> But in that case you are not worst than a pure HTML button, or are you?

With a vanilla HTML button like `<button id=myButton>OK</button>`, you can change the (declaratively or programatically set) label by setting `myButton.textContent` or `myButton.innerHTML`.   So, it's easy to do programatically, and perhaps more importantly, presumably works with MVC frameworks.

Unfortunately, this strategy doesn't work with a deliteful Button widget because the label is in a sub-node, not the root node.   You could tell the application to set `myButton.containerNode.textContent`, but I don't think that will work well with MVC frameworks.

Ideally we would support setting `textContent` or `innerHTML` directly but it seems too difficult to implement.  Thus, mapping `label` to `containerNode.textContent` seems like a good compromise.

**Comment by wkeese** [comment]
> I guess I'll come back to my initial point which is why making label such a specific case? We have not other property fo the widget that is filled automatically by its content? 

Actually, we have many cases like that:

* Button and subclasses currently work like that (initial textContent --> `label` property)
* delite/Store subclasses (initial textContent --> `source` property)
* Combobox (initial child `<d-list>` node --> `list` property)

Regardless, the point of #290 is not to change existing behavior, but to enhance Button to support markup like:

```xml
<d-button>
   <d-progress-indicator></d-progress-indicator>
   refreshing...
</d-button>
```

If there's a consensus to change the existing behavior, we could do that separately.
